client/beefy/src/worker.rs

```diff
@@ -345,11 +350,24 @@ where
 			// update best GRANDPA finalized block we have seen
 			self.best_grandpa_block_header = header.clone();
 
-			// Check for and enqueue potential new session.
-			if let Some(new_validator_set) = find_authorities_change::<B>(header) {
-				self.init_session_at(new_validator_set, *header.number());
-				// TODO: when adding SYNC protocol, fire up a request for justification for this
-				// mandatory block here.
+			// Check all (newly) finalized blocks for new session(s).
+			let backend = self.backend.clone();
+			for header in notification
+				.tree_route
+				.iter()
+				.map(|hash| {
+					backend
+						.blockchain()
+						.expect_header(BlockId::hash(*hash))
+						.expect("just finalized block should be available; qed.")
```

Comment:
I'd also avoid calling expect here - there are db issues. concurrency, caches, mocked backends, ... they may cause their own issues. Otoh BEEFY worker may actually stop working because of that, so I can't say that logging + ignoring an error is any better :)

Reply:
yes, that is my thinking as well, if this errors out on a session boundary the voter is dead in the water anyway.

This way at least you can see the beefy-gadget task crashing and restart it..